feat(adapter-mariadb): support passing a pool instance

aqrln opened the pull request:
Allow `@prisma/adapter-mariadb` to accept a pre-configured pool instance, matching what `@prisma/adapter-pg` already supports.

```ts
const pool = mariadb.createPool({ ... })
const adapter = new PrismaMariaDb(pool)
```

## Pool ownership

The adapter does not end a pool it did not create. For an externally provided pool, `dispose()` is a no-op and the caller stays responsible for ending it. Pass `disposeExternalPool: true` to have the adapter call `pool.end()` on dispose instead.

With `disposeExternalPool: true`, `connect()` may be called only once. The adapter ends the pool it was given and cannot create a replacement from it, so a second adapter would end up sharing (and then ending) a pool the first one is still using. When the adapter owns the pool it created, it simply builds a new one from the connection settings, so repeated `connect()` calls are fine.

## Caveat: `prepareCacheLength`

When the adapter creates its own pool, it defaults `prepareCacheLength` to 0 to work around the prepared-statement leak in #29364. It cannot apply that default to a pool it was handed, so callers passing their own pool should set `prepareCacheLength: 0` themselves unless they want the driver's prepared-statement cache.

Ref: https://github.com/prisma/prisma/issues/26885
Ref: https://github.com/prisma/prisma/pull/27062

aqrln committed:
feat(adapter-mariadb): support passing a pool instance

aqrln commented on packages/adapter-mariadb/src/mariadb.ts:
```diff
@@ -170,19 +172,38 @@ export class PrismaMariaDbAdapterFactory implements SqlDriverAdapterFactory {
 
   #capabilities?: Capabilities
   #config: mariadb.PoolConfig | string
+  #externalPool: mariadb.Pool | null
```
This needs to be a tagged union, we either have a config or a pool (since we can't retrieve a config from a pool with `mariadb` driver). We should also check that `connect` can't be called more than once if an external pool is used and `disposeExternalPool` is true.

tensordreams commented on packages/adapter-mariadb/src/mariadb.ts:
```diff
@@ -218,53 +224,93 @@ export class PrismaMariaDbAdapterFactory implements SqlDriverAdapterFactory {
   async connect(): Promise<PrismaMariaDbAdapter> {
-    let pool: mariadb.Pool
-    try {
-      pool = mariadb.createPool(this.#config)
-    } catch (error) {
-      // We match on an error which is known to leak the connection string and replace it with
-      // a custom error message.
-      // The error might change in a future version of the driver, but this is covered in a
-      // test, which checks for credential leakage regardless of the exact error message.
-      if (error instanceof Error && error.message.startsWith('error parsing connection string')) {
-        throw new Error(
-          "error parsing connection string, format must be 'mariadb://[<user>[:<password>]@]<host>[:<port>]/[<db>[?<opt1>=<value1>[&<opt2>=<value2>]]]'",
-        )
-      }
-      throw error
+    if (this.#externalPoolDisposed) {
+      throw new Error(
+        'Cannot call connect() again after an external pool was disposed by the adapter. ' +
+          'Either let the adapter manage the pool by passing connection settings instead, or set ' +
+          '`disposeExternalPool: false` and end the pool yourself once you no longer need the adapter.',
+      )
     }
```
Addressed in f24037a. The post-disposal boolean is replaced by a claim staked at `connect()` time, before any `await`:

```ts
if (poolOrConfig.type === 'pool' && ownsPool) {
  if (this.#externalPoolClaimed) throw new Error(...)
  this.#externalPoolClaimed = true
}
```

This closes both windows you identified. A second `connect()` is now rejected while the first adapter is still live, while its `dispose()` is in flight, and when `pool.end()` rejects — the release callback no longer mutates factory state at all, so a failed shutdown can't leave the pool re-claimable.

This also matches the reviewer request from the original round: `connect()` is single-shot whenever an external pool is combined with `disposeExternalPool: true`.

Covered by three tests, each verified to fail without the guard.

tensordreams commented on packages/adapter-mariadb/src/mariadb.ts:
```diff
@@ -218,53 +224,93 @@ export class PrismaMariaDbAdapterFactory implements SqlDriverAdapterFactory {
+
     if (this.#capabilities === undefined) {
       this.#capabilities = await getCapabilities(pool)
     }
```
Skipping this one: `getCapabilities()` cannot reject, so the proposed `catch` would be dead code.

Its entire body — including `rows[0][0]` and `inferCapabilities()` — is inside a `try` whose `catch` swallows the error and returns a default:

```ts
async function getCapabilities(pool: mariadb.Pool) {
  try {
    const rows = await pool.query({ sql: `SELECT VERSION()`, rowsAsArray: true })
    const version = rows[0][0]
    ...
    return capabilities
  } catch (e) {
    debug(`${tag} Error while checking capabilities: %O`, e)
    return { supportsRelationJoins: false }
  }
}
```

An unreachable database therefore yields `{ supportsRelationJoins: false }` rather than a rejection, and `connect()` returns an adapter that owns the pool and can release it through `dispose()`. There is no leak path here.

The other failure mode in this area — `createPool()` throwing on a malformed connection string — happens before any pool exists, so there is nothing to clean up there either.

tensordreams committed:
fix(adapter-mariadb): don't log connection string parse errors

Node's ERR_INVALID_URL carries the rejected string in error.input, so logging
the error object put the connection string's credentials into the debug
history. That history is recorded whether or not DEBUG is set, and getLogs()
embeds it in the error reports users paste into GitHub issues.

Log a generic message instead, and extend the credential sanitization suite to
cover the debug path alongside the thrown error.

tensordreams commented on packages/adapter-mariadb/src/mariadb.ts:
```diff
@@ -218,53 +224,100 @@ export class PrismaMariaDbAdapterFactory implements SqlDriverAdapterFactory {
+      }
+      return rewriteConnectionString(url).toString()
+    } catch (error) {
+      debug('Error parsing connection string: %O', error)
```
Valid, and worse than described — fixed in c6dd1f8.

Confirmed the leak vector:

```
> new URL('mariadb://user:super_secret_pw_12345@:3306/db')
TypeError: Invalid URL {
  code: 'ERR_INVALID_URL',
  input: 'mariadb://user:super_secret_pw_12345@:3306/db'
}
```

The part worth flagging: this is not confined to users who set `DEBUG`. In `@prisma/debug`, `argsHistory.push()` runs *unconditionally*, before the `enabled` check — and `getLogs()` reads that history, which `getErrorMessageWithLink.ts` embeds in the "please report this" error text users paste into GitHub issues. So a malformed connection string could put a password into a public issue with debug output never enabled.

`normalizeConfig()` now logs a generic message and drops the error object. Added a `getLogs()`-based regression test to `credentials.test.ts`, next to the existing thrown-error one, following the precedent in `adapter-pg`; verified it fails when the old `debug(..., error)` call is restored.

Note for the record that this line came from #29392 and predates this PR — this branch only relocated it into `normalizeConfig()`. Fixing it here since the PR already touches the line and the fix is one line, but it applies to `main` independently.
